### PR TITLE
Cleanup `MachineThread`

### PR DESCRIPTION
**Pull request by joshuacolvin0**
* Finish pending thread if trying to start new machine thread
* Avoid dereferencing nullptr
* make `Machine` and `MachineThread` destructor virtual

### CODE REVIEW

File: packages/arb-avm-cpp/data_storage/src/arbcore.cpp [hunk 1]
```diff
@@ -1452,6 +1458,11 @@ void ArbCore::printCoreThreadBacktrace() {
 bool ArbCore::reorgIfInvalidMachine(uint32_t& thread_failure_count,
                                     uint256_t& next_checkpoint_gas,
                                     ValueCache& cache) {
+    if (core_machine->status() == MachineThread::MACHINE_RUNNING) {
+        // Just continue
+        return true;
+    }
```
Comment (hkalodner) [inline packages/arb-avm-cpp/data_storage/src/arbcore.cpp:1464]: I'm not sure why we need this?

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-avm-cpp/avm/src/machinethread.cpp [hunk 2]
```diff
@@ -24,6 +24,7 @@ MachineThread::machine_status_enum MachineThread::status() {
 
 bool MachineThread::runMachine(MachineExecutionConfig config,
                                bool asynchronous) {
+    finishThread();
```
[the remark this answers is not in the document]
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/avm/src/machinethread.cpp]: `runMachine` should only return false on error condition.  Updated to simply return true if asynchronous and still running.  Code expects machine to be in a stopped state after `runMachine` executes if asynchronous is false, so finishThread is appropriate.

File: packages/arb-avm-cpp/data_storage/include/data_storage/executioncursor.hpp [hunk 3]
```diff
@@ -72,24 +72,29 @@ class ExecutionCursor {
...
+                std::get<std::unique_ptr<Machine>>(machine)->abort();
+            }
         }
     }
 
     bool isAborted() { return is_aborted.load(); }
 
     [[nodiscard]] std::optional<uint256_t> machineHash() const {
-        if (std::holds_alternative<std::unique_ptr<Machine>>(machine)) {
+        if (std::holds_alternative<std::unique_ptr<Machine>>(machine) &&
```
[the remark this answers is not in the document]
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/data_storage/include/data_storage/executioncursor.hpp:87]: :+1:

File: packages/arb-avm-cpp/data_storage/include/data_storage/executioncursor.hpp [hunk 4]
```diff
@@ -72,24 +72,29 @@ class ExecutionCursor {
...
+            std::get<std::unique_ptr<Machine>>(machine) != nullptr) {
             return std::get<std::unique_ptr<Machine>>(machine)->hash();
         } else {
             return std::get<MachineStateKeys>(machine).machineHash();
         }
     }
 
     [[nodiscard]] const MachineOutput& getOutput() const {
-        if (std::holds_alternative<std::unique_ptr<Machine>>(machine)) {
+        if (std::holds_alternative<std::unique_ptr<Machine>>(machine) &&
```
[the remark this answers is not in the document]
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/data_storage/include/data_storage/executioncursor.hpp:96]: :+1:

File: packages/arb-avm-cpp/data_storage/src/arbcore.cpp (hunk 1 above)
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/data_storage/src/arbcore.cpp:1464]: Fixed to call `core_machine->finishThread()` instead.  `getReorgData` is not thread safe, so want to make certain that machine is done running.

File: packages/arb-avm-cpp/avm/src/machinethread.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/avm/src/machinethread.cpp]: removed completely

[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-avm-cpp/avm/src/machinethread.cpp]: `runMachine` isn't considered thread safe and is only called from the core thread so I don't think the synchronous case matters. Personally I'd also view it as an API violation to call `runMachine` when the machine isn't in the `MACHINE_NONE` state. I think there's a large chance that if `runMachine` is called when the machine is already running there's a significant error in the code

File: packages/arb-avm-cpp/avm/src/machinethread.cpp [hunk 5]
```diff
@@ -48,6 +56,7 @@ bool MachineThread::runMachine(MachineExecutionConfig config,
 }
 
 bool MachineThread::continueRunningMachine(bool asynchronous) {
+    finishThread();
```
Comment (hkalodner) [inline packages/arb-avm-cpp/avm/src/machinethread.cpp:59]: Whenever we resolve the `runMachine` discussion, the same will apply here I think

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-avm-cpp/avm/src/machinethread.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/avm/src/machinethread.cpp]: treating as fatal error

File: packages/arb-avm-cpp/avm/src/machinethread.cpp (hunk 5 above)
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/avm/src/machinethread.cpp]: treating as fatal error

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**